### PR TITLE
feat(agent): add support for function-based agents and buffer editor

### PR DESCRIPTION
**Pull request by SDGLBL**
- **.gitignore**: Exclude macOS .DS_Store files.
- **AGENTS.md**: Updated to include information on function-based agents and buffer editor implementation.
- **README.md**: Documented the new Buffer Editor agent and distinctions between agent types.
- **lua/codecompanion/agents.lua**: Added execute function for Buffer Editor agent.
- **lua/codecompanion/config.lua**: Configured Buffer Editor agent.
- **lua/codecompanion/helpers/buffers.lua**: Implemented buffer content formatting and buffer info retrieval.
- **lua/codecompanion/strategies.lua**: Included context in agent strategy.
- **lua/codecompanion/utils/util.lua**: Added function to get active buffers.

The introduction of function agents significantly enhances the plugin's capabilities and flexibility. This type of agent can perform more complex and powerful operations, going beyond the default chat interface interaction model. 

CleanShot 2024-07-12 at 02 18 16

### CODE REVIEW

**Comment by olimorris** [comment]
This looks awesome!!! I'll try it out tomorrow and feedback.

**Comment by olimorris** [comment]
2024-07-12 09_19_51 - Safari@2x

A few tweaks I'd like to make to the docs and config. Could you give me access to your fork?

File: lua/codecompanion/adapters/anthropic.lua [hunk 1]
```diff
@@ -108,7 +108,7 @@ return {
       local output = {}
 
       -- Skip the event messages
-      if type(data) == "string" and string.sub(data, 1, 6) == "event:" then
+      if type(data) == "string" and (string.sub(data, 1, 6) == "event:" or data == "data: [DONE]") then
```
Comment (olimorris) [inline lua/codecompanion/adapters/anthropic.lua:111]: Can we remove this change as I don't believe it's part of the Claude API response.

Source: https://docs.anthropic.com/en/api/messages-streaming#raw-http-stream-response

**Comment by olimorris** [comment]
Thanks so much for this. I've just played around with it now and it's awesome! I was hoping that someone would implement this type of feature in the plugin as it's the next logical step of allowing the LLM to implement its code in the editor.

Your idea to use `CDATA` in the XML prompt is fantastic. I've played around asking the agent to do minor changes and it works well. For more complex asks like "refactor this file" it's less able but we can think about how we could implement that in the future.

Just a couple of minor requests to the config and README.

File: lua/codecompanion/adapters/anthropic.lua (hunk 1 above)
Comment (SDGLBL) [reply lua/codecompanion/adapters/anthropic.lua:111]: ok

**Comment by olimorris** [comment]
Thanks for giving me access. I find working on PRs to be impossible at times:

2024-07-12 10_22_47 - WezTerm@2x

I'm happy to merge the PR as it is now and make the changes later. Anything new you want to add before I merge?

**Comment by SDGLBL** [comment]
> Thanks for giving me access. I find working on PRs to be impossible at times:
> 
> 
> 
> 2024-07-12 10_22_47 - WezTerm@2x
> 
> 
> 
> I'm happy to merge the PR as it is now and make the changes later. Anything new you want to add before I merge?

If that's the case, let's merge it first. I can submit new pull requests in the coming days to make some minor bug fixes. One obvious bug I've found so far is that there's an issue with creating new files in the buffer editor.​​​​​​​​​​​​​​​​

**Pull request merged**